api-v2/src/interceptors/transform-response-interceptor.ts

```diff
@@ -0,0 +1,32 @@
+import {
+  CallHandler,
+  ExecutionContext,
+  Injectable,
+  NestInterceptor,
+} from '@nestjs/common';
+import { Observable } from 'rxjs';
+import { map } from 'rxjs/operators';
+
+export interface ServerResponse<T> {
+  statusCode: number;
+  data: T;
+}
+
+/**
+ * Transforms all outgoing data into the ServerResponse shape above.
+ */
+@Injectable()
+export class TransformResponseInterceptor<T>
```

Comment:
oop my bad the comment is outdated, i'll change it. but essentially it transforms all outgoing data into the `ServerDataResponse` type which essentially standardizes the shape of what our server responds with.

I did this cause i found it kinda odd to have our responses simply be the actual data. For instance the student get endpoint returns the student itself so a client would essentially do `response.catalogYear` etc. At the same time I didn't want to manually go and return a specific response shape from each controller method. So this global interceptor takes the data returned from controllers and transforms it into a standardized recognizable shape -
```
{
  data: T,
  statusCode: number
}
```
This meshes kinda well with the `ServerErrorResponse` type(which is what nest returns if we throw an HttpError). So our frontend can safely expect either a ServerErrorResponse or a ServerDataResponse. We can then use a type guard to handle both cases on the frontend with full type safety. Essentially the frontend would be forced to handle the error case before proceeding to handle the data case.

All the above types are at the top of the common/types file!

Does this make sense? Would love to know your thoughts on this, it's just a design decision I made so is definitely arguable! Would also

Reply:
Makes sense, but I think including the http code in the body is redundant.

on the frontend, we can explicitly handle all non GET request responses anyway, so I think it's fine to not include. Thoughts?

Reply:
yeap no reason to handle the statusCode so I can strip that off!

Reply:
and then i guess after that, there's no point in wrapping `data` in an object, because axios does this already, so i think it would be good to remove the entire wrapper